tensilelite/rocisa/rocisa/include/code.hpp

```diff
@@ -0,0 +1,1613 @@
+            : Item(name)
+            , ref(std::nullopt)
+            , value(value)
+            , offset(offset)
+            , format(format)
+        {
+        }
+
+        ValueSet(const std::string& name, const std::string& ref, int offset = 0, int format = 0)
+            : Item(name)
+            , ref(ref)
+            , value(std::nullopt)
+            , offset(offset)
+            , format(format)
+        {
+        }
+
+        std::string toString() const override
+        {
+            std::string t = ".set " + name + ", ";
+            if(ref)
+            {
+                if(format == -1)
+                {
+                    t += ref.value();
+                }
+                else
+                {
+                    t += ref.value() + "+" + std::to_string(offset);
+                }
+            }
+            else if(value)
+            {
+                if(format == -1)
+                {
+                    t += std::to_string(value.value());
+                }
+                else if(format == 0)
+                {
+                    t += std::to_string(value.value() + offset);
+                }
+                else if(format == 1)
+                {
+                    t += "0x" + toHex(value.value() + offset);
+                }
+            }
+            t += "\n";
+            return t;
+        }
+
+    private:
+        std::string toHex(int num) const
+        {
+            std::stringstream ss;
+            ss << std::hex << num;
+            return ss.str();
+        }
+    };
+
+    struct RegSet : public ValueSet
+    {
+        std::string regType;
+
+        RegSet(const std::string& regType, const std::string& name, int value, int offset = 0)
+            : ValueSet(name, value, offset)
+            , regType(regType)
+        {
+        }
+
+        RegSet(const std::string& regType,
+               const std::string& name,
+               const std::string& value,
+               int                offset = 0)
+            : ValueSet(name, value, offset)
+            , regType(regType)
+        {
+        }
+    };
+
+    inline std::string field_desc(const std::string& field_name, int value, int bits = 0)
+    {
+        std::string bits_str = (bits > 0) ? " (" + std::to_string(bits) + "b)" : "";
+        return field_name + bits_str + ": " + std::to_string(value);
+    }
+
+    struct BitfieldUnion
+    {
+        virtual std::string toString() const
+        {
+            return "0x" + toHex(value);
+        }
+
+        virtual int getValue() const
+        {
+            return value;
+        }
+
+        virtual std::string fields_desc() const = 0;
+        virtual std::string desc() const        = 0;
+
+    protected:
+        int value;
+
+        std::string toHex(int num) const
+        {
+            std::stringstream ss;
+            ss << std::hex << num;
+            return ss.str();
+        }
+    };
+
+    union SrdUpperFields9XX
+    {
+        struct
```

Comment:
Just a comment. Will these fields be copied one by one or with a single 32b copy when used to init the buffer descriptor? If doing a single copy, I'm not 100% sure the ordering of the bit fields is guaranteed to be respected by the compiler.

Reply:
single 23b